dfget/core/api/download_api.go

```diff
@@ -67,12 +66,71 @@ func (d *downloadAPI) Download(ip string, port int, req *DownloadRequest, timeou
 		}
 	}
 
-	var url string
-	if strings.Contains(req.Path, "://") {
+	var (
+		url      string
+		rangeStr string
+	)
+	if isFromSource(req) {
+		rangeStr = getRealRange(req.PieceRange, headers[config.StrRange])
 		url = req.Path
 	} else {
+		rangeStr = httputils.ConstructRangeStr(req.PieceRange)
 		url = fmt.Sprintf("http://%s:%d%s", ip, port, req.Path)
 	}
+	headers[config.StrRange] = rangeStr
 
 	return httputils.HTTPGetTimeout(url, headers, timeout)
 }
+
+func isFromSource(req *DownloadRequest) bool {
+	return req != nil && strings.Contains(req.Path, "://")
+}
+
+// getRealRange
+// pieceRange: "start-end"
+// rangeHeaderValue: "bytes=sourceStart-sourceEnd"
+// return: "bytes=realStart-realEnd"
+func getRealRange(pieceRange string, rangeHeaderValue string) string {
+	if rangeHeaderValue == "" {
+		return httputils.ConstructRangeStr(pieceRange)
+	}
+	rangeEle := strings.Split(rangeHeaderValue, "=")
+	if len(rangeEle) != 2 {
+		return httputils.ConstructRangeStr(pieceRange)
+	}
+
+	lower, upper, err := parseRange(rangeEle[1])
+	if err != nil {
+		return httputils.ConstructRangeStr(pieceRange)
+	}
+	start, end, err := parseRange(pieceRange)
+	if err != nil {
+		return httputils.ConstructRangeStr(pieceRange)
+	}
+
+	realStart := start + lower
+	realEnd := end + lower
+	if realEnd > upper {
+		realEnd = upper
+	}
+	return httputils.ConstructRangeStr(fmt.Sprintf("%d-%d", realStart, realEnd))
+}
+
+func parseRange(rangeStr string) (int64, int64, error) {
```

Comment:
https://github.com/dragonflyoss/Dragonfly/blob/master/supernode/util/range_util.go#L54

It's a repeated function that we extract it as common. WDYT?

Reply:
done